src/simulator/particle/LinkedCellContainer.cpp

```diff
@@ -0,0 +1,372 @@
+#include "LinkedCellContainer.h"
+#include <cmath>
+
+size_t Cell::size() const { return particles.size(); }
+
+ParticlePointer Cell::operator[](size_t index) { return particles[index]; }
+
+LinkedCellContainer::LinkedCellContainer(
+    std::initializer_list<double> domain_size, double r_cutoff,
+    std::initializer_list<double> left_corner_coordinates,
+    const DomainBoundaryConditions &boundary_conditions)
+    : domain_size_{domain_size}, r_cutoff_{r_cutoff},
+      left_corner_coordinates{left_corner_coordinates}, x{0}, y{0}, z{0},
+      boundary_conditions_{boundary_conditions}, ParticleContainer{} {
+  if (domain_size.size() != 3 && domain_size.size() != 2) {
+    throw std::invalid_argument("Domain size must have 2 or 3 elements");
+  }
+
+  double remainder_x = std::fmod(domain_size_[0], r_cutoff);
+  double remainder_y = std::fmod(domain_size_[1], r_cutoff);
+  double remainder_z =
+      domain_size.size() == 3 ? std::fmod(domain_size_[2], r_cutoff) : 0.0;
+  if (std::abs(remainder_x) > DIVISION_TOLERANCE)
+    extend_x = true;
+
+  if (std::abs(remainder_y) > DIVISION_TOLERANCE)
+    extend_y = true;
+
+  if (std::abs(remainder_z) > DIVISION_TOLERANCE)
+    extend_z = true;
+
+  x = static_cast<size_t>(domain_size_[0] / r_cutoff) + (extend_x ? 1 : 0);
+  y = static_cast<size_t>(domain_size_[1] / r_cutoff) + (extend_y ? 1 : 0);
+  z = domain_size.size() == 3
+          ? (static_cast<size_t>((domain_size_[2] / r_cutoff)) +
+             (extend_z ? 1 : 0))
+          : 1;
```

Comment:
If I understand correctly, you are creating an extra small cell at the end of the domain, if the domain size is not exactly divisible by the cutoff. However, I see this as a potential problem for several reasons:
- time step is set in a way that particle doesn't cross an entire cell in one step. So, now with smaller cell, there might be cases when you need small timestep than specified in the worksheet. And depending on how small the cell is, this can be really small and create problems in the simulation. (this is just for you to be aware, but also in general you want to be able to run the simulation with as large a timestep as possible)
- for reflection with ghost particle: if cell is too small then reflection should already happen one cell before as particle is too close to boundary. If you only consider particle in these small boundary cells, they might be too close and create much higher repulsive force and cause simulation to explode?

some other approaches:
- just use a slightly bigger uniformly-sized cells
- just make the last cell a little bigger